clients/src/main/java/org/apache/kafka/clients/consumer/CooperativeStickyAssignor.java

```diff
@@ -0,0 +1,120 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+import static org.apache.kafka.clients.consumer.CooperativeStickyAssignor.CooperativeUserData.decode;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.kafka.common.protocol.types.Field;
+import org.apache.kafka.common.protocol.types.Schema;
+import org.apache.kafka.common.protocol.types.SchemaException;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.protocol.types.Type;
+
+/**
+ * A cooperative version of the {@link StickyAssignor StickyAssignor}. This follows the same (sticky)
+ * assignment logic as {@link StickyAssignor StickyAssignor} but allows for cooperative rebalancing. To turn on
+ * cooperative incremental rebalancing you must set all your consumers to use this {@code PartitionAssignor},
+ * or implement a custom one that returns {@code RebalanceProtocol.COOPERATIVE} in
+ * {@link CooperativeStickyAssignor#supportedProtocols supportedProtocols()}.
+ * <p>
+ * IMPORTANT: if upgrading from 2.3 or earlier, you must follow a specific upgrade path in order to safely turn on
+ * cooperative rebalancing. See the upgrade-guide for details
+ */
+public class CooperativeStickyAssignor extends StickyAssignor {
+
+    @Override
+    public List<RebalanceProtocol> supportedProtocols() {
+        return Arrays.asList(RebalanceProtocol.EAGER, RebalanceProtocol.COOPERATIVE);
+    }
+
+    @Override
+    public String name() {
+        return "cooperative-sticky";
+    }
+
+    @Override
+    Subscription buildSubscription(Set<String> topics) {
+        return new Subscription(
+            new ArrayList<>(topics), new CooperativeUserData(generation()).encode(), memberAssignment());
+    }
+
+    @Override
+    StickyUserData subscriptionToStickyUserData(Subscription subscription) {
+        CooperativeUserData userData = decode(subscription.userData());
+        return new StickyUserData(subscription.ownedPartitions(), Optional.of(userData.generation));
+    }
+
+    static final class CooperativeUserData {
+
+        private static final short COOPERATIVE_USERDATA_V0 = 0;
+        static final short LATEST_VERSION = COOPERATIVE_USERDATA_V0;
```

Comment:
Not in the scope of this PR but: in the "future works" as we encode the version into the join-group request as well, maybe it's better to move this and the versions in StreamsAssignor into a centralized place?

EDIT: actually nvm, I think it's better keeping them separated since not all assignors would be versioned, and there's less sense to enforce it anyways.